src/Psalm/Type.php

```diff
@@ -600,6 +600,9 @@ public static function intersectUnionTypes(
                         );
 
                         if (null !== $intersection_atomic) {
+                            if ($intersection_atomic instanceof TNever) {
+                                return new Union([$intersection_atomic]);
+                            }
```

Comment:
I'm not quite sure this is correct. What if you have something like [this](https://psalm.dev/r/332b1a4764) and we end up calling `intersectUnionTypes` with `(Foo&A)|(Bar&B)` and `Bar&B`, I think this would call `intersectAcomicTypes` on `Foo&A` and `Bar&B`, get `TNever` and return `never`, when it should actually skip past the `never` result from the first iteration and return the `Bar&B` result when intersecting `Bar&B` and `Bar&B`.

Reply:
Ah yes, I see what you mean. I'll iterate.

Reply:
Okay, this doesn't seem to be an issue - but I'm not entirely sure why. I've added a few extra tests, and can't get it to fall over. I can only assume that types derived from docblocks don't call this code?

Reply:
Yeah, I'm having trouble actually finding a way to call `intersectUnionTypes` as well... It might be easiest to just add a normal unit test that creates the types and calls it directly.

Reply:
I think it's because `Type::intersectAtomicTypes` is slightly broken in that it's not checking for impossible intersections between two classes, so when you intersect `Foo&A` and `Bar&B` you actually get `Foo&A&Bar&B` instead of `never`. I'd say just fix it without having a test.

Reply:
Here's the test case I stepped through to figure it out:

```php
class Foo {}
class Bar {}
interface A {}
interface B {}
class FooA extends Foo implements A {}
class BarB extends Bar implements B {}

interface FooBarFactoryInterface
{
    /** @return (Foo&A)|(Bar&B) */
    public function create(): object;
}

interface FactoryInterface
{
    public function create(): object;
}

abstract class AbstractFooAFactory
{
    public function create(): Foo&A
    {
        return new FooA();
    }
}

class FooAFactory extends AbstractFooAFactory implements FooBarFactoryInterface, FactoryInterface {}

function getFooAFactory(): AbstractFooAFactory&FooBarFactoryInterface&FactoryInterface
{
    return new FooAFactory();
}

$factory = getFooAFactory();
$foobar = $factory->create();
```

It's pretty brittle though and there are a lot of unrelated things that could change and cause it to not test what it's supposed to be testing. If/when this is fixed I think it's worth having a separate test that just calls `intersectUnionTypes` directly.